ruma-serde/src/canonical_json/value.rs

```diff
@@ -0,0 +1,244 @@
+use std::{
+    collections::BTreeMap,
+    convert::{TryFrom, TryInto},
+    fmt,
+};
+
+use js_int::Int;
+use serde::de::Error;
+use serde_json::Value as JsonValue;
+
+#[derive(Clone, Eq, PartialEq)]
+pub enum CanonicalJsonValue {
+    /// Represents a JSON null value.
+    ///
+    /// ```
+    /// # use serde_json::json;
+    /// # use std::convert::TryInto;
+    /// # use ruma_serde::CanonicalJsonValue;
+    /// let v: CanonicalJsonValue = json!(null).try_into().unwrap();
+    /// ```
+    Null,
+
+    /// Represents a JSON boolean.
+    ///
+    /// ```
+    /// # use serde_json::json;
+    /// # use std::convert::TryInto;
+    /// # use ruma_serde::CanonicalJsonValue;
+    /// let v: CanonicalJsonValue = json!(true).try_into().unwrap();
+    /// ```
+    Bool(bool),
+
+    /// Represents a JSON integer.
+    ///
+    /// ```
+    /// # use serde_json::json;
+    /// # use std::convert::TryInto;
+    /// # use ruma_serde::CanonicalJsonValue;
+    /// let v: CanonicalJsonValue = json!(12).try_into().unwrap();
+    /// ```
+    Integer(Int),
+
+    /// Represents a JSON string.
+    ///
+    /// ```
+    /// # use serde_json::json;
+    /// # use std::convert::TryInto;
+    /// # use ruma_serde::CanonicalJsonValue;
+    /// let v: CanonicalJsonValue = json!("a string").try_into().unwrap();
+    /// ```
+    String(String),
+
+    /// Represents a JSON array.
+    ///
+    /// ```
+    /// # use serde_json::json;
+    /// # use std::convert::TryInto;
+    /// # use ruma_serde::CanonicalJsonValue;
+    /// let v: CanonicalJsonValue = json!(["an", "array"]).try_into().unwrap();
+    /// ```
+    Array(Vec<CanonicalJsonValue>),
+
+    /// Represents a JSON object.
+    ///
+    /// The map is backed by a BTreeMap to guarantee the sorting of keys.
+    ///
+    /// ```
+    /// # use serde_json::json;
+    /// # use std::convert::TryInto;
+    /// # use ruma_serde::CanonicalJsonValue;
+    /// let v: CanonicalJsonValue = json!({ "an": "object" }).try_into().unwrap();
+    /// ```
+    Object(BTreeMap<String, CanonicalJsonValue>),
+}
+
+impl CanonicalJsonValue {
+    /// Returns a canonical JSON string according to Matrix specification.
+    ///
+    /// The method should be preferred over `serde_json::to_string` since it
+    /// checks the size of the canonical string. Matrix canonical JSON enforces
+    /// a size limit of less than 65,535 when sending PDU's for the server-server protocol.
+    pub fn to_canonical_string(&self) -> Result<String, serde_json::Error> {
+        Ok(serde_json::to_string(self).and_then(|s| {
+            if s.len() > 65_535 {
+                Err(serde_json::Error::custom("TOO LARGE"))
+            } else {
+                Ok(s)
+            }
+        })?)
+    }
+}
+
+impl fmt::Debug for CanonicalJsonValue {
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            CanonicalJsonValue::Null => formatter.debug_tuple("Null").finish(),
```

Comment:
```suggestion
            Self::Null => formatter.debug_tuple("Null").finish(),
```

here and in other places.